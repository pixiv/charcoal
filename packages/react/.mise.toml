[tools]
pnpm = "10.17.1"
