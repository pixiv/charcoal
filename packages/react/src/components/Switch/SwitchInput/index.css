.charcoal-switch-input {
  cursor: pointer;
  appearance: none;
  display: inline-flex;
  position: relative;
  box-sizing: border-box;
  width: 28px;
  border: 2px solid transparent;

  transition-property: background-color, box-shadow;
  transition-duration: 0.2s;

  outline: none;
  border-radius: 16px;
  height: 16px;
  margin: 0;
  background-color: var(--charcoal-color-container-neutral-default);
}

.charcoal-switch-input:disabled,
.charcoal-switch-input[readonly] {
  cursor: default;
}

.charcoal-switch-input::after {
  content: '';
  position: absolute;
  display: block;
  top: 0;
  left: 0;
  width: 12px;
  height: 12px;
  transform: translateX(0);
  transition: transform 0.2s;
  border-radius: 1024px;
  background-color: var(--charcoal-color-icon-on-primary-default);
}

.charcoal-switch-input:checked::after {
  transform: translateX(12px);
  transition: transform 0.2s;
}

.charcoal-switch-input:checked {
  background-color: var(--charcoal-color-container-primary-default);
}

.charcoal-switch-input:not(:disabled):hover {
  background-color: var(--charcoal-color-container-neutral-hover);
}

.charcoal-switch-input:not(:disabled):active {
  background-color: var(--charcoal-color-container-neutral-press);
}

.charcoal-switch-input:not(:disabled):focus {
  outline: none;
  box-shadow: var(--charcoal-color-custom-focus-effect);
}
.charcoal-switch-input:not(:disabled):focus-visible {
  box-shadow: var(--charcoal-color-custom-focus-effect);
}
.charcoal-switch-input:not(:disabled):focus:not(:focus-visible) {
  box-shadow: none;
}

.charcoal-switch-input:not(:disabled)::after:hover {
  background-color: var(--charcoal-color-icon-on-primary-hover);
}

.charcoal-switch-input:not(:disabled)::after:active {
  background-color: var(--charcoal-color-icon-on-primary-press);
}

.charcoal-switch-input:not(:disabled):checked:hover {
  background-color: var(--charcoal-color-container-primary-hover);
}

.charcoal-switch-input:not(:disabled):checked:active {
  background-color: var(--charcoal-color-container-primary-press);
}
