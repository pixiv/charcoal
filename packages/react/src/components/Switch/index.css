.charcoal-switch__label {
  display: inline-grid;
  grid-template-columns: auto 1fr;
  align-items: center;
  cursor: pointer;
  outline: 0;
  gap: 4px;
}

.charcoal-switch__label[aria-disabled='true'] {
  cursor: default;
  opacity: 0.32;
}

.charcoal-switch__label_div {
  font-size: 14px;
  line-height: 22px;
  color: var(--charcoal-color-text-default);
}
