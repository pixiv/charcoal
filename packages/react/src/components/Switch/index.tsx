import { useSwitch } from '@react-aria/switch'
import type { AriaSwitchProps } from '@react-types/switch'
import React, { useRef, useMemo } from 'react'
import { useToggleState } from 'react-stately'
import styled from 'styled-components'
import { theme } from '../../styled'
import { disabledSelector } from '@charcoal-ui/utils'

export type SwitchProps = {
  name: string
  className?: string
  value?: string
  checked?: boolean
  disabled?: boolean
  onChange(checked: boolean): void
} & (
  | // children か label は片方が必須
  {
      children: React.ReactNode
    }
  | {
      label: string
    }
)

export default function SwitchCheckbox(props: SwitchProps) {
  const { disabled, className } = props

  const ariaSwitchProps: AriaSwitchProps = useMemo(
    () => ({
      ...props,

      // children がいない場合は aria-label をつけないといけない
      'aria-label': 'children' in props ? undefined : props.label,
      isDisabled: props.disabled,
      isSelected: props.checked,
    }),
    [props]
  )

  const state = useToggleState(ariaSwitchProps)
  const ref = useRef<HTMLInputElement>(null)
  const {
    inputProps: { className: _className, type: _type, ...rest },
  } = useSwitch(ariaSwitchProps, state, ref)

  return (
    <Label className={className} aria-disabled={disabled}>
      <SwitchInput {...rest} ref={ref} />
      {'children' in props ? (
        // eslint-disable-next-line react/destructuring-assignment
        <LabelInner>{props.children}</LabelInner>
      ) : undefined}
    </Label>
  )
}

const Label = styled.label`
  display: inline-grid;
  grid-template-columns: auto 1fr;
  align-items: center;
  cursor: pointer;
  outline: 0;

  ${theme((o) => o.disabled)}

  ${disabledSelector} {
    cursor: default;
  }
`

const LabelInner = styled.div`
  ${theme((o) => [
    o.typography(14).preserveHalfLeading,
    o.font.text2,
    o.margin.left(4),
  ])}
`

const SwitchInput = styled.input.attrs({
  type: 'checkbox',
})`
  &[type='checkbox'] {
    appearance: none;
    display: inline-flex;
    position: relative;
    box-sizing: border-box;
    width: 28px;
    border: 2px solid transparent;
    transition: box-shadow 0.2s, background-color 0.2s;
    cursor: inherit;
    ${theme((o) => [
      o.borderRadius(16),
      o.height.px(16),
      o.bg.text4.hover.press,
      o.outline.default.focus,
      o.margin.all(0),
    ])}

    &::after {
      content: '';
      position: absolute;
      display: block;
      top: 0;
      left: 0;
      width: 12px;
      height: 12px;
      transform: translateX(0);
      transition: transform 0.2s;
      ${theme((o) => [o.bg.text5.hover.press, o.borderRadius('oval')])}
    }

    &:checked {
      /* FIXME: o.outline.default.focus の transition に o.bg.brand の transition が打ち消されてしまう */
      transition: all 0.2s !important;
      ${theme((o) => o.bg.brand.hover.press)}

      &::after {
        transform: translateX(12px);
      }
    }
  }
`
