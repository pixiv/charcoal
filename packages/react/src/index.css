:root {
  /* charcoal custom css variables */
  --charcoal-color-custom-focus-effect: 0 0 0 4px rgba(0, 150, 250, 0.32);
  --charcoal-color-custom-error-effect: 0 0 0 4px rgba(255, 43, 0, 0.32);
}
