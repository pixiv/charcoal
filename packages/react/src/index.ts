export {
  default as ComponentAbstraction,
  useComponentAbstraction,
  type LinkProps,
} from './core/ComponentAbstraction'
export { SSRProvider } from './core/SSRProvider'
export { default as Button, type ButtonProps } from './components/Button'
export {
  default as Clickable,
  type ClickableProps,
  type ClickableElement,
} from './components/Clickable'
export {
  default as IconButton,
  type IconButtonProps,
} from './components/IconButton'
export {
  default as Radio,
  type RadioProps,
  RadioGroup,
  type RadioGroupProps,
} from './components/Radio'
export {
  default as Select,
  type SelectProps,
  SelectGroup,
  type SelectGroupProps,
} from './components/Select'
export { default as Switch, type SwitchProps } from './components/Switch'
export {
  default as TextField,
  type TextFieldProps,
  type SingleLineTextFieldProps,
  type MultiLineTextFieldProps,
} from './components/TextField'
export { default as Icon, type IconProps } from './components/Icon'
export { default as Modal, type ModalProps } from './components/Modal'
export {
  default as LoadingSpinner,
  LoadingSpinnerIcon,
} from './components/LoadingSpinner'
export {
  default as SegmentedControl,
  type SegmentedControlProps,
} from './components/SegmentedControl'
export { default as Checkbox } from './components/Checkbox'
export { default as TagItem, type TagItemProps } from "./components/TagItem"
