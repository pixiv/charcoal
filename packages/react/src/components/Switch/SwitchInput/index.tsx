import './index.css'

import { forwardRef, ForwardRefExoticComponent, RefAttributes, useCallback } from 'react'
import { useClassNames } from '../../../_lib/useClassNames'

export type CharcoalSwitchProps = {
  onChange?: (checked: boolean) => void
}

type InputProps = React.HTMLProps<HTMLInputElement>

export type SwitchInputProps = CharcoalSwitchProps &
  Omit<InputProps, keyof CharcoalSwitchProps | 'ref'>

const SwitchInput: ForwardRefExoticComponent<Omit<SwitchInputProps, "children"> & RefAttributes<HTMLInputElement>> = forwardRef<
  HTMLInputElement,
  Omit<SwitchInputProps, 'children'>
>(function SwitchInput({ onChange, className, ...props }, ref) {
  const handleChange = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      const el = e.currentTarget
      onChange?.(el.checked)
    },
    [onChange]
  )

  const classNames = useClassNames('charcoal-switch-input', className)
  return (
    <input
      ref={ref}
      className={classNames}
      type="checkbox"
      onChange={handleChange}
      {...props}
    />
  )
})

export default SwitchInput
