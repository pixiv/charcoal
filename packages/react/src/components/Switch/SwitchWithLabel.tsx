import * as React from 'react'
import { useClassNames } from '../../_lib/useClassNames'

export const SwitchWithLabel: React.NamedExoticComponent<{
  children: React.ReactNode
  className?: string
  disabled?: boolean
  id?: string
  input: React.ReactNode
}> = React.memo(function SwitchWithLabel({
  children,
  className,
  disabled,
  id,
  input,
}: {
  children: React.ReactNode
  className?: string
  disabled?: boolean
  id?: string
  input: React.ReactNode
}) {
  const classNames = useClassNames('charcoal-switch__label', className)
  return (
    <label htmlFor={id} className={classNames} aria-disabled={disabled}>
      {input}
      <div className="charcoal-switch__label_div">{children}</div>
    </label>
  )
})
