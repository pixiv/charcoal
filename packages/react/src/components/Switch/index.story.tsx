import { useState } from 'react'
import Switch from '.'
import { Meta, StoryObj } from '@storybook/react'

export default {
  title: 'Switch',
  component: Switch,
  parameters: {
    layout: 'centered',
  },
} as Meta<typeof Switch>

export const Default: StoryObj<typeof Switch> = {
  render: function Render(args) {
    const [checked, setChecked] = useState(false)
    return (
      <div>
        <Switch
          {...args}
          name="name"
          onChange={(v) => {
            setChecked(v)
          }}
          checked={args.checked ?? checked}
        />
      </div>
    )
  },
}

export const Checked: StoryObj<typeof Switch> = {
  render: function Render() {
    const [checked, setChecked] = useState(true)
    return (
      <div>
        <Switch
          name="name"
          label="checked"
          onChange={(v) => {
            setChecked(v)
          }}
          checked={checked}
        />
      </div>
    )
  },
}

export const Label: StoryObj<typeof Switch> = {
  render: function Render() {
    const [checked, setChecked] = useState(false)
    return (
      <div>
        <Switch name="label" checked={checked} onChange={setChecked}>
          Label
        </Switch>
      </div>
    )
  },
}

export const Disabled: StoryObj<typeof Switch> = {
  render: function Render() {
    const [checked, setChecked] = useState(false)
    return (
      <div>
        <Switch
          name="disabled"
          checked={checked}
          onChange={setChecked}
          disabled
        >
          Label
        </Switch>
      </div>
    )
  },
}
