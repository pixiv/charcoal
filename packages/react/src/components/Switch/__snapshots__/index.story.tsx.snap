// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Labelled smoke-test 1`] = `
.c0 {
  display: inline-grid;
  grid-template-columns: auto 1fr;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  outline: 0;
}

.c0:disabled,
.c0[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c0:active > input {
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c2 {
  font-size: 14px;
  line-height: 22px;
  color: var(--charcoal-text2);
  margin-left: 4px;
}

.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  box-sizing: border-box;
  width: 28px;
  border: 2px solid transparent;
  -webkit-transition-property: background-color,box-shadow;
  transition-property: background-color,box-shadow;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  cursor: inherit;
  outline: none;
  border-radius: 16px;
  height: 16px;
  margin: 0;
  background-color: var(--charcoal-text4);
}

.c1:hover {
  background-color: var(--charcoal-text4-hover);
}

.c1:active {
  background-color: var(--charcoal-text4-press);
}

.c1:focus {
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c1::after {
  content: '';
  position: absolute;
  display: block;
  top: 0;
  left: 0;
  width: 12px;
  height: 12px;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  border-radius: 1024px;
  background-color: var(--charcoal-text5);
}

.c1::after:hover {
  background-color: var(--charcoal-text5-hover);
}

.c1::after:active {
  background-color: var(--charcoal-text5-press);
}

.c1:checked {
  background-color: var(--charcoal-brand);
}

.c1:checked:hover {
  background-color: var(--charcoal-brand-hover);
}

.c1:checked:active {
  background-color: var(--charcoal-brand-press);
}

.c1:checked::after {
  -webkit-transform: translateX(12px);
  -ms-transform: translateX(12px);
  transform: translateX(12px);
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
}

<div data-dark="false">
  <div>
    <label class="c0"
           aria-disabled="false"
    >
      <input name="name"
             role="switch"
             type="checkbox"
             class="c1"
      >
      <div class="c2">
        選択する
      </div>
    </label>
  </div>
</div>
`;

exports[`Switch Playground smoke-test 1`] = `
.c0 {
  display: inline-grid;
  grid-template-columns: auto 1fr;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  outline: 0;
}

.c0:disabled,
.c0[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c0:active > input {
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c2 {
  font-size: 14px;
  line-height: 22px;
  color: var(--charcoal-text2);
  margin-left: 4px;
}

.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  box-sizing: border-box;
  width: 28px;
  border: 2px solid transparent;
  -webkit-transition-property: background-color,box-shadow;
  transition-property: background-color,box-shadow;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  cursor: inherit;
  outline: none;
  border-radius: 16px;
  height: 16px;
  margin: 0;
  background-color: var(--charcoal-text4);
}

.c1:hover {
  background-color: var(--charcoal-text4-hover);
}

.c1:active {
  background-color: var(--charcoal-text4-press);
}

.c1:focus {
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c1::after {
  content: '';
  position: absolute;
  display: block;
  top: 0;
  left: 0;
  width: 12px;
  height: 12px;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  border-radius: 1024px;
  background-color: var(--charcoal-text5);
}

.c1::after:hover {
  background-color: var(--charcoal-text5-hover);
}

.c1::after:active {
  background-color: var(--charcoal-text5-press);
}

.c1:checked {
  background-color: var(--charcoal-brand);
}

.c1:checked:hover {
  background-color: var(--charcoal-brand-hover);
}

.c1:checked:active {
  background-color: var(--charcoal-brand-press);
}

.c1:checked::after {
  -webkit-transform: translateX(12px);
  -ms-transform: translateX(12px);
  transform: translateX(12px);
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
}

<div data-dark="false">
  <div>
    <label class="c0">
      <input name="name"
             role="switch"
             type="checkbox"
             class="c1"
      >
      <div class="c2">
        選択する
      </div>
    </label>
  </div>
</div>
`;

exports[`Switch Unlabelled smoke-test 1`] = `
.c0 {
  display: inline-grid;
  grid-template-columns: auto 1fr;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  outline: 0;
}

.c0:disabled,
.c0[aria-disabled]:not([aria-disabled='false']) {
  opacity: 0.32;
  cursor: default;
}

.c0:active > input {
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
  box-sizing: border-box;
  width: 28px;
  border: 2px solid transparent;
  -webkit-transition-property: background-color,box-shadow;
  transition-property: background-color,box-shadow;
  -webkit-transition-duration: 0.2s;
  transition-duration: 0.2s;
  cursor: inherit;
  outline: none;
  border-radius: 16px;
  height: 16px;
  margin: 0;
  background-color: var(--charcoal-text4);
}

.c1:hover {
  background-color: var(--charcoal-text4-hover);
}

.c1:active {
  background-color: var(--charcoal-text4-press);
}

.c1:focus {
  box-shadow: 0 0 0 4px rgba(0,150,250,0.32);
}

.c1::after {
  content: '';
  position: absolute;
  display: block;
  top: 0;
  left: 0;
  width: 12px;
  height: 12px;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
  border-radius: 1024px;
  background-color: var(--charcoal-text5);
}

.c1::after:hover {
  background-color: var(--charcoal-text5-hover);
}

.c1::after:active {
  background-color: var(--charcoal-text5-press);
}

.c1:checked {
  background-color: var(--charcoal-brand);
}

.c1:checked:hover {
  background-color: var(--charcoal-brand-hover);
}

.c1:checked:active {
  background-color: var(--charcoal-brand-press);
}

.c1:checked::after {
  -webkit-transform: translateX(12px);
  -ms-transform: translateX(12px);
  transform: translateX(12px);
  -webkit-transition: -webkit-transform 0.2s;
  -webkit-transition: transform 0.2s;
  transition: transform 0.2s;
}

<div data-dark="false">
  <div>
    <label class="c0"
           aria-disabled="false"
    >
      <input aria-label="label"
             name="name"
             role="switch"
             type="checkbox"
             class="c1"
      >
    </label>
  </div>
</div>
`;
